### PR TITLE
[VSTS] Move away from Jenkins to VSTS pipelines for xamarin-macios.

### PR DESCRIPTION
**Pull request by mandel-macaque**
This PR (that will be squash) moves all the CI from Jenkins to VSTS pipelines. There are some small details to add in later PRs:

* Add API diffs. #10299
* Add simple links for the nugets and pkgs. #10298
* Add the nugets to the feed. #10300
* Clean old jenkins folder.

### CODE REVIEW

**Comment by mandel-macaque** [comment]
@spouliot @rolfbjarne looks like vsts webhook does not set a pending status. Will add one when it starts from the pipeline itself after the merge is done.

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:92]: This seems a bit simpler:

```suggestion
- bash: |
    set -x
    sudo rm -Rf /Developer/MonoTouch
    sudo rm -Rf /Library/Frameworks/Xamarin.iOS.framework
    sudo rm -Rf /Library/Frameworks/Xamarin.Mac.framework
```

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:127]: Labels can be set for all builds, not only PR builds, so this logic should always be executed.

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:315]: Why are PR builds and CI builds signed differently?

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:408]: This will be green if the failure-stamp file wasn't created, you need to return RC:

```suggestion
    fi
    
    exit $RC
```

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:389]: Make sure errors aren't ignored:

```suggestion
    set -x
    set -e
    echo "Running tests on $AGENT_NAME"
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tools/devops/automation/build-pipeline.yml
Comment (spouliot) [inline tools/devops/automation/build-pipeline.yml:177]: questions
* how do the pools handle Catalina and BigSur ?
* how do the pools handle 11.0 and 11.1 ? all updated asap vs different pools ?
* any pool for beta macOS, e.g. upcoming 11.2 ?

File: tools/devops/automation/scripts/System.psm1
Comment (spouliot) [reply tools/devops/automation/scripts/System.psm1:208]: > # Xcode 10.2.1 is currently used by Binding Tools for Swift # 

has not BTfS updated to something more recent this fall ?

File: tools/devops/provision-objective-sharpie.csx
Comment (spouliot) [inline tools/devops/provision-objective-sharpie.csx:3]: what's calling this file ?

File: tools/devops/automation/scripts/System.psm1
Comment (mandel-macaque) [reply tools/devops/automation/scripts/System.psm1:208]: @rolfbjarne if you looked, i'm using your bash script, I'll remove this function since it is not needed.

File: .gitignore
Comment (mandel-macaque) [reply .gitignore:16]: yes, in this PR, I can rename it, but is the templates related to the building of the pacakges.

File: tools/devops/automation/build-pipeline.yml
Comment (mandel-macaque) [reply tools/devops/automation/build-pipeline.yml:73]: Thos are the target branches, yes, should we do it for PRs to all branches? @mjbond-msft can that be done?

File: tools/devops/automation/build-pipeline.yml
Comment (mandel-macaque) [reply tools/devops/automation/build-pipeline.yml:177]: ATM we have a pool for each, but I updated the agent to be able to filter using the Aget.OS, @mjbond-msft was going to create a pool with all the needed devices, in that case, we woud use a single pool for 11. and 11. Etc

@mjbond-msft can we have beta bots on those pools with unstable OS? I don't think it would be an issue as long as the agent works.

File: tools/devops/automation/build-pipeline.yml
Comment (mandel-macaque) [reply tools/devops/automation/build-pipeline.yml:179]: I have TODO it :)

File: tools/devops/automation/scripts/bash/productsign.sh
Comment (mandel-macaque) [reply tools/devops/automation/scripts/bash/productsign.sh:1]: It is the same script. But was moved outside the jenkins folder.

File: tools/devops/automation/templates/agent-pool-selector.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/agent-pool-selector.yml:57]: @mjbond-msft You are the person to answer this question.

File: tools/devops/automation/templates/mac-tests.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/mac-tests.yml:34]: Yes, there is another branch with that work nearly completed.

File: tools/devops/provision-vs.csx
Comment (mandel-macaque) [reply tools/devops/provision-vs.csx:28]: @mjbond-msft ?

File: tools/devops/automation/scripts/bash/productsign.sh
Comment (rolfbjarne) [reply tools/devops/automation/scripts/bash/productsign.sh:1]: The old script is still there, which means git will show this as an added file (instead of a file move).

File: tools/devops/provision-objective-sharpie.csx
Comment (mandel-macaque) [reply tools/devops/provision-objective-sharpie.csx:3]: There is a better way to approach this, I'll add it, we can use make to get the sharpie version and add it to the provisioning

File: tools/devops/automation/build-pipeline.yml
Comment (mjbond-msft) [reply tools/devops/automation/build-pipeline.yml:177]: We've already have a pool `VSEng-Xamarin-RedmondMacTextPool-iOS`, which currently has one Big Sur machine. We can add machines as needed to that pool

File: tools/devops/automation/scripts/bash/productsign.sh
Comment (mandel-macaque) [reply tools/devops/automation/scripts/bash/productsign.sh:1]: I'll step on it

File: tools/devops/automation/templates/mac-tests.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/mac-tests.yml:2]: we do not, old comment, will remove it.

File: tools/devops/automation/templates/devices/build.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/devices/build.yml:109]: Will apply.. we found a bug in github suggestions
Screenshot 2020-12-16 at 11 09 30

 Will do it manually :/

File: tools/devops/automation/templates/packages/build.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/packages/build.yml:306]: Honestly, I don't know I followed https://github.com/xamarin/xamarin-macios/blob/main/jenkins/Jenkinsfile#L568

File: tools/devops/automation/templates/agent-pool-selector.yml
Comment (mjbond-msft) [reply tools/devops/automation/templates/agent-pool-selector.yml]: Tightened up the comments to describe more precisely what trigger types & branches each pool (trusted & untrusted) serves

Comment (rolfbjarne) [reply tools/devops/automation/templates/agent-pool-selector.yml]: What's the reason for the split? According to the comments it seems to be because the `main` and `d16-*` branches are more trusted in some way, but there's no reason to trust any other (non-fork) branch more or less than `main`/ `d16-*` as far as I know.

File: tools/devops/provision-vs.csx
Comment (mjbond-msft) [reply tools/devops/provision-vs.csx]: Agree we can avoid the callout to the .ps1 file in this case and absorb the code directly into the CSX script. However, given the current solution works (i.e., don't fix what isn't broken), I recommend addressing this after the PR merges.  I've created a follow item to resolve this post-merge: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1258715

File: tools/devops/automation/templates/agent-pool-selector.yml
Comment (mjbond-msft) [reply tools/devops/automation/templates/agent-pool-selector.yml]: The intent of the split is to isolate hardware pools so that the CI builds for reviewed code (i.e., code that has completed PR review & has been merged to staging (main & release branches -- aka branches with policies associated) is handled by dedicated "trusted" hardware; whereas, all other jobs (commits to feature/topic branches) and PR merge commits are handled by an untrusted pool.  Helps better protect machines in the trusted pool that produce bits that ship in the product from being compromised. It may be worth having a discussion on what makes the most sense for the iOS team from this perspective

**Review by tj-devel709 — APPROVED** [review]

File: tools/devops/automation/templates/packages/build.yml
Comment (mandel-macaque) [reply tools/devops/automation/templates/packages/build.yml:394]: I've realised, we don't want that, that will fail the build, but we want to continue to be able to report the results, else we will get out.

File: tools/devops/automation/templates/agent-pool-selector.yml
Comment (rolfbjarne) [reply tools/devops/automation/templates/agent-pool-selector.yml]: So if I understand correctly, the intent is to protect the bots that produce our releases from _accidental_ damage by any of us (by creating a branch to test things and that ends up breaking the bot somehow). If that's the case, it should be explained in a comment, because my first thought was that it was trying to protect something from malicious use, and this split does not help in that scenario at all.

File: tools/devops/automation/templates/mac-tests.yml
Comment (rolfbjarne) [reply tools/devops/automation/templates/mac-tests.yml:33]: Is this filed somewhere so that we don't forget?

File: tools/devops/automation/build-pipeline.yml
Comment (rolfbjarne) [reply tools/devops/automation/build-pipeline.yml:73]: Yes, we need PR builds for all PRs, it doesn't matter the target branch.

File: tools/devops/automation/build-pipeline.yml
Comment (rolfbjarne) [reply tools/devops/automation/build-pipeline.yml:179]: Is it filed somewhere (else)?

File: tools/devops/automation/templates/packages/build.yml
Comment (rolfbjarne) [inline tools/devops/automation/templates/packages/build.yml:197]: If we're building on machines that has access to maccore, even for PR builds, we might as well always pass `--enable-xamarin`

File: .gitignore
Comment (rolfbjarne) [reply .gitignore:16]: We create `packages` directories that shouldn't be in `git` in other places, so with this change as it stands, those other directories need to be found and the corresponding `.gitignore` files added/updated to include `packages` to avoid unnecessary dirty working directories.

File: tools/devops/automation/templates/packages/build.yml
Comment (akoeplinger) [reply tools/devops/automation/templates/packages/build.yml:197]: what stops someone from filing a PR that .zip's maccore and uploads it somewhere? the public build shouldn't have access to private bits if you want to trigger the pipeline automatically.

Comment (rolfbjarne) [reply tools/devops/automation/templates/packages/build.yml:197]: We wouldn't trigger PRs from people without write access automatically, they'd need to be triggered manually.

Comment (akoeplinger) [reply tools/devops/automation/templates/packages/build.yml:197]: That's a bit annoying for contributors but OK :)

**Review by dalexsoto — COMMENTED** [review]
Looking good so far 👍

**Review by spouliot — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Comment by mandel-macaque** [comment]
Merging, I have created several issues to track the missing work:

* https://github.com/xamarin/maccore/issues/2349 
* https://github.com/xamarin/maccore/issues/2350
* https://github.com/xamarin/xamarin-macios/issues/10299
* https://github.com/xamarin/xamarin-macios/issues/10298
* https://github.com/xamarin/xamarin-macios/issues/10300
* https://github.com/xamarin/maccore/issues/2351

File: .gitignore
Comment (mandel-macaque) [reply .gitignore:16]: Issue https://github.com/xamarin/maccore/issues/2359 (has already a PR).

**Pull request merged**
